### PR TITLE
Add policy evaluation task stage to the CLI

### PR DESCRIPTION
**Pull request by mrinalirao**
**Note: Builds on #32302**

This PR is the second PR for adding OPA support to the CLI.
This PR adds the policy evaluation task stage to the runTaskStage method.

Testing
Necessary unit tests added

Links
https://hashicorp.atlassian.net/browse/TF-1456
Designs : https://www.figma.com/file/euexPgt2fRF2VXUwDOgTMP/Terraform-%7C-Policy-Evaluation-%5BWIP%5D?node-id=1642%3A63889

Screen Shot 2022-11-29 at 1 04 16 pm

### CODE REVIEW

**Comment by mrinalirao** [comment]
Task stage overrides local tests:
Screen Shot 2022-11-30 at 10 44 20 am

Screen Shot 2022-11-30 at 10 45 44 am

**Comment by glennsarti** [comment]
I think we might need some work on errored policies.  I can't really test cancelled ones

In TFC UI

image


In TF CLI

image

**Comment by mrinalirao** [comment]
Good catch. Now returning an error if the task stage fails, errors or is canceled.

Screen Shot 2022-12-01 at 12 07 47 pm

**Comment by glennsarti** [comment]
Do you think we need to output the policies?

When a run task fails you get:

image

But when an OPA errors:
image

File: internal/cloud/backend_taskStages.go [hunk 1]
```diff
@@ -30,3 +52,142 @@ func (b *Cloud) runTaskStages(ctx context.Context, client *tfe.Client, runId str
...
+				}
+				if !cont {
+					continue
+				}
+				// cont is true and we must continue to poll
+				if msg != nil {
+					output.OutputElapsed(*msg, len(*msg)) // Up to 2 digits are allowed by the max message allocation
+				}
+				return true, nil
+			}
```
Comment (brandonc) [inline internal/cloud/backend_taskStages.go:151]: 💅 There is a lot of duplication here which makes this method harder to follow than it needs to be. Specifically this part, duplicated 3 times. I wonder if you could encapsulate this in a new type that could be a little more explicit about the overall control flow here. It's hard to follow because the inner loop can print a status message or the summarizer can but it's unclear where exactly the responsibility lies.

```
for _, s := range summarizers {
	cont, msg, err := s.Summarize(ctx, output, stage)
	if err != nil {
		errs.Append(err)
		break
	}
	if !cont {
		continue
	}

	// cont is true and we must continue to poll
	if msg != nil {
		output.OutputElapsed(*msg, len(*msg)) // Up to 2 digits are allowed by the max message allocation
	}
	return true, nil
}
```

**Review by brandonc — APPROVED** [review]
I've suggested some refactors but the smoke test worked well. The control flow between polling and summarizing feels complicated and it was hard for me to determine if the unit tests account for all possibilities from the API. Despite all that, I don't suppose my feedback is blocking! Thanks for working to integrate some of these existing pieces together

**Review by brandonc — APPROVED** [review]

File: internal/cloud/tfe_client_mock.go [hunk 2]
```diff
@@ -27,6 +27,8 @@ type MockClient struct {
 	CostEstimates         *MockCostEstimates
 	Organizations         *MockOrganizations
 	Plans                 *MockPlans
+	PolicySetOutcomes     *MockPolicySetOutcomes
```
[the remark this answers is not in the document]
Comment (brandonc) [reply internal/cloud/tfe_client_mock.go:30]: Let me know if you'd like to me to merge anyway!

[the remark this answers is not in the document]
Comment (mrinalirao) [reply internal/cloud/tfe_client_mock.go:30]: PolicySetOutcome is the output of the OPA policy evaluation, this is added so we can mock the List and Read func which are being used to test the policy evaluation. Let me know if that clarifies things :)
https://github.com/hashicorp/go-tfe/blob/main/policy_evaluation.go#L120
